src/Common/src/System/Runtime/InteropServices/StringBuffer.cs

```diff
@@ -0,0 +1,315 @@
+            if (sourceIndex < 0 || sourceIndex > source.Length) throw new ArgumentOutOfRangeException("sourceIndex");
+            if (count < 0) count = source.Length - sourceIndex;
+            if (source.Length < sourceIndex + count) throw new ArgumentOutOfRangeException("count");
+
+            if (count == 0) return;
+            ulong lastIndex = bufferIndex + (ulong)count;
+            if (_length < lastIndex) Length = lastIndex;
+
+            fixed (char* content = source)
+            {
+                Buffer.MemoryCopy(
+                    source: content + sourceIndex,
+                    destination: CharPointer + bufferIndex,
+                    destinationSizeInBytes: checked((long)(ByteCapacity - (bufferIndex * sizeof(byte)))),
+                    sourceBytesToCopy: count * sizeof(char));
+            }
+        }
+
+        /// <summary>
+        /// Trim the specified values from the end of the buffer. If nothing is specified, nothing is trimmed.
+        /// </summary>
+        public unsafe void TrimEnd(char[] values)
+        {
+            if (values == null || values.Length == 0 || _length == 0) return;
+
+            char* end = CharPointer + _length - 1;
+
+            while (Array.IndexOf(values, *end) >= 0)
+            {
+                Length--;
+                end--;
+            }
+        }
+
+        /// <summary>
+        /// String representation of the entire buffer. If the buffer is larger than the maximum size string (int.MaxValue) will truncate.
+        /// </summary>
+        public unsafe override string ToString()
+        {
+            if (_length == 0) return string.Empty;
+            return new string(CharPointer, startIndex: 0, length: _length > int.MaxValue ? int.MaxValue : checked((int)_length));
+        }
+
+        /// <summary>
+        /// Get the given substring in the buffer.
+        /// </summary>
+        /// <param name="count">Count of characters to take, or remaining characters from <paramref name="startIndex"/> if -1.</param>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// Thrown if <paramref name="startIndex"/> or <paramref name="count"/> are outside the range of the buffer's length
+        /// or count is greater than the maximum string size (int.MaxValue).
+        /// </exception>
+        public unsafe string Substring(ulong startIndex, int count = -1)
+        {
+            if (_length > 0 && startIndex > _length - 1) throw new ArgumentOutOfRangeException("startIndex");
+            if (count < -1) throw new ArgumentOutOfRangeException("count");
+
+            ulong realCount = count == -1 ? _length - startIndex : (ulong)count;
+            if (realCount > int.MaxValue || startIndex + realCount > _length) throw new ArgumentOutOfRangeException("count");
+            if (realCount == 0) return string.Empty;
+
+            return new string(value: CharPointer + startIndex, startIndex: 0, length: checked((int)realCount));
```

Comment:
Why add startIndex to the CharPointer rather than just passing it as the startIndex argument? e.g.
```C#
return new string(CharPointer, startIndex, (int)realCount);
```
Also not clear to me why the checked is necessary: a few lines earlier we verified that realCount is <= int.MaxValue.